File: frontend/src/lib/tmdb.ts
import axios from 'axios'

const TMDB_API_KEY = (import.meta as any).env?.VITE_TMDB_API_KEY || '28ad717f54f5cf84820b45205b4521df'
const TMDB_READ_TOKEN = (import.meta as any).env?.VITE_TMDB_READ_TOKEN
const TMDB_BASE = 'https://api.themoviedb.org/3'
export const TMDB_IMG = 'https://image.tmdb.org/t/p'

// Use API key in URL params (more reliable than Bearer token for browser)
const client = axios.create({
  baseURL: TMDB_BASE,
  params: { api_key: TMDB_API_KEY },
  headers: { Accept: 'application/json' }
})

// Debug: log which auth method is being used
console.log('TMDB Auth: API Key', `(${TMDB_API_KEY.substring(0, 8)}...)`)

export type TmdbMovieLite = {
  id: number
  title: string
  year?: string
  poster?: string
}

export type TmdbMovieFull = {
  id: number
  title: string
  year?: string
  runtime?: number
  genres?: string
  overview?: string
  poster?: string
  rating?: number
}

export async function getTrendingMovies(): Promise<TmdbMovieLite[]> {
  try {
    // Use /movie/popular endpoint (same as user's working example)
    const { data } = await client.get('/movie/popular')
    console.log('TMDB popular response:', data)
    if (!data || !data.results) {
      throw new Error('Invalid response from TMDB API')
    }
    return data.results.map((m: any) => ({
      id: m.id,
      title: m.title ?? m.name ?? '',
      year: m.release_date ? String(m.release_date).slice(0, 4) : undefined,
      poster: m.poster_path ? `${TMDB_IMG}/w500${m.poster_path}` : undefined,
    }))
  } catch (error: any) {
    console.error('TMDB popular error:', error.response?.data || error.message)
    const errorMsg = error.response?.data?.status_message || error.message || 'Failed to fetch movies'
    throw new Error(errorMsg)
  }
}

export async function searchMoviesTmdb(query: string): Promise<TmdbMovieLite[]> {
  try {
    const { data } = await client.get('/search/movie', { params: { query, language: 'en-US', include_adult: false } })
    console.log('TMDB search response:', data)
    return (data?.results ?? []).map((m: any) => ({
      id: m.id,
      title: m.title ?? m.name ?? '',
      year: m.release_date ? String(m.release_date).slice(0, 4) : undefined,
      poster: m.poster_path ? `${TMDB_IMG}/w342${m.poster_path}` : undefined,
    }))
  } catch (error: any) {
    console.error('TMDB search error:', error.response?.data || error.message)
    return []
  }
}

export async function getMovieByIdTmdb(id: string | number): Promise<TmdbMovieFull | null> {
  try {
    const { data } = await client.get(`/movie/${id}`, { params: { language: 'en-US' } })
    if (!data) return null
    return {
      id: data.id,
      title: data.title,
      year: data.release_date ? String(data.release_date).slice(0, 4) : undefined,
      runtime: data.runtime,
      genres: Array.isArray(data.genres) ? data.genres.map((g: any) => g.name).join(', ') : undefined,
      overview: data.overview,
      poster: data.poster_path ? `${TMDB_IMG}/w500${data.poster_path}` : undefined,
      rating: typeof data.vote_average === 'number' ? data.vote_average : undefined,
    }
  } catch (error: any) {
    console.error('TMDB movie details error:', error.response?.data || error.message)
    return null
  }
}

File: backend/app/services/tmdb.py
import requests
from typing import Optional, List, Dict, Any
from ..core.config import get_settings

settings = get_settings()
TMDB_BASE = "https://api.themoviedb.org/3"
TMDB_IMG = "https://image.tmdb.org/t/p"


def get_tmdb_headers() -> Dict[str, str]:
    """Get headers for TMDB API requests"""
    headers = {"Accept": "application/json"}
    if settings.tmdb_read_token:
        headers["Authorization"] = f"Bearer {settings.tmdb_read_token}"
    return headers


def get_tmdb_params() -> Dict[str, str]:
    """Get query params for TMDB API requests"""
    params = {}
    if settings.tmdb_api_key and not settings.tmdb_read_token:
        params["api_key"] = settings.tmdb_api_key
    return params


def get_popular_movies() -> List[Dict[str, Any]]:
    """Fetch popular movies from TMDB"""
    try:
        url = f"{TMDB_BASE}/movie/popular"
        response = requests.get(
            url,
            headers=get_tmdb_headers(),
            params=get_tmdb_params(),
            timeout=10
        )
        response.raise_for_status()
        data = response.json()
        return data.get("results", [])
    except Exception as e:
        print(f"TMDB API error: {e}")
        return []


def search_movies(query: str) -> List[Dict[str, Any]]:
    """Search movies on TMDB"""
    try:
        url = f"{TMDB_BASE}/search/movie"
        params = get_tmdb_params()
        params.update({"query": query, "language": "en-US", "include_adult": False})
        response = requests.get(
            url,
            headers=get_tmdb_headers(),
            params=params,
            timeout=10
        )
        response.raise_for_status()
        data = response.json()
        return data.get("results", [])
    except Exception as e:
        print(f"TMDB search error: {e}")
        return []


def get_movie_details(movie_id: int) -> Optional[Dict[str, Any]]:
    """Get movie details by ID from TMDB"""
    try:
        url = f"{TMDB_BASE}/movie/{movie_id}"
        params = get_tmdb_params()
        params["language"] = "en-US"
        response = requests.get(
            url,
            headers=get_tmdb_headers(),
            params=params,
            timeout=10
        )
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f"TMDB movie details error: {e}")
        return None

